README.md
[![Les Laboratoires Ruby](https://invidget.switchblade.xyz/4P7XcmbDnt)](https://discord.gg/4P7XcmbDnt)

# RubyDate

Welcome to your new gem! In this directory, you'll find the files you need to be able to package up your Ruby library into a gem. Put your Ruby code in the file `lib/rubyDate`. To experiment with that code, run `bin/console` for an interactive prompt.

This gem is a date manager gem. This is useful to use.

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'rubyDate'
```

And then execute:

    $ bundle install

Or install it yourself as:

    $ gem install rubyDate

## Usage
 ###Formatter
This is still in wip, but there a simple usage

```ruby
require "yaml"

file = YAML::load(File.open("lib/data/formats.yml"))

puts RubyDate::Formatter.new(:fr).format(file[:fr])
```
 ###Duration
A useful example for the duration class
```ruby
duration = RubyDate::Duration.new(:fr, 60 * 60 * 24 * 365)
puts duration.format("", nil, { :object => false, :array => false })
```
An example using the viewer
```ruby
duration = RubyDate::Duration.new(:fr, 60 * 60 * 24 * 365)
puts duration.format("", :hours, { :object => false, :array => false })
```
And an example returning the duration in an array
```ruby
duration = RubyDate::Duration.new(:fr, 60 * 60 * 24 * 365)
puts duration.format("", :hours, { :object => false, :array => true })
```

## Development

After checking out the repo, run `bin/setup` to install dependencies. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and the created tag, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/[USERNAME]/rubyDate.

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).


CHANGELOG.md
## [Unreleased]

## [0.1.0] - 2021-04-10

- Initial release


lib/rubyDate/formatter.rb
require "date"
require "time"
require "yaml"
require_relative "constants"

module RubyDate
  class Formatter
    attr_reader :date, :format_data
    attr_accessor :locale, :formats_path, :langs_path

    def initialize(locale, date = Time.new, formats_path = "lib/data/formatter/formats.yml", langs_path = "lib/data/formatter/langs.yml")
      @locale, @date, @formats_path, @langs_path = locale, date, formats_path, langs_path
      @format_data = YAML::load(File.open(@formats_path))
      @lang_data = YAML::load(File.open(@langs_path))
    end

    def format(format)
      date = @date.to_datetime.strftime(format)
      case @locale
      when :en
        date
      else
        if @format_data[@locale]
          @lang_data[:months][@locale].each_key do |key|
            next unless date.include?(key)
            date[key] = @lang_data[:months][@locale][key]
          end
          @lang_data[:days][@locale].each_key do |key|
            next unless date.include?(key)
            date[key] = @lang_data[:days][@locale][key]
          end
          date
        else raise "Bad locale format #{@locale}"; end
      end
    end
  end
end

lib/rubyDate/duration.rb
require "time"
require "date"
require "yaml"
require_relative "formatter"

module RubyDate
  class Duration
    attr_reader :relative_format_path, :relative_format_data, :duration
    attr_accessor :time, :lang
    def initialize(lang, time = 0, relative_format_path = "lib/data/relativeFormat.yml")
      @lang, @time, @relative_format_path = lang, time, relative_format_path
      @relative_format_data = YAML::load(File.open(@relative_format_path))
    end
    def format(format, viewer = nil, options = { :object => false, :array => false })
      timestamp = (Time.new + @time) - Time.new
      seconds = Time.strptime("#{timestamp}", "%s").to_i # seconds from epoch
      minutes, seconds_left, hours, minutes_left, days, hours_left, days_left, years = 0, 0, 0, 0, 0, 0, 0, 0
      if seconds >= 60
        minutes, seconds_left = seconds.divmod(60)
        if minutes >= 60
          hours, minutes_left = minutes.divmod(60)
          if hours >= 24
            days, hours_left = hours.divmod(24)
            if days >= 365
              years = days.divmod(365)[0]
            end
          end
        end
      end
      duration = Hash[
        :seconds => Hash[
          :total => seconds,
          :left => seconds_left
        ],
        :minutes => Hash[
          :total => minutes,
          :left => minutes_left
        ],
        :hours => Hash[
          :total => hours,
          :left => hours_left
        ],
        :days => Hash[
          :total => days,
          :left => days_left
        ],
        :years => Hash[
          :total => years
        ]
      ]
      if options[:object]
        duration
      else
        if viewer.nil?
          formats = {
            :seconds => seconds_left > 1 ? :ss : :s,
            :minutes => minutes_left > 1 ? :mm : :m,
            :hours => hours_left > 1 ? :hh : :h,
            :days => days_left > 1 ? :dd : :d,
            :years => years > 1 ? :yy : :y
          }
          strings = {}
          formats.each do |key, value|
            next if duration[key][:left] === 0 and key != :seconds or duration[key][:total] === 0
            unless @relative_format_data[value][@lang]; raise "No lang found : #{@lang}"; end
            strings[key] = "#{@relative_format_data[value][@lang]}"
          end
          format_date_duration = []
          strings.each_value do |value|
            format_date_duration << "#{value}"
          end
          format_date_duration.each do |duration_string|
            # Replace by seconds
            if duration_string.include? "[%ss]"
              duration_string["[%ss]"] = seconds_left.to_s
              next
            end
            # Replace by minutes
            if duration_string.include? "[%mm]"
              duration_string["[%mm]"] = minutes_left.to_s
              next
            end
            # Replace by hours
            if duration_string.include? "[%hh]"
              duration_string["[%hh]"] = hours_left.to_s
              next
            end
            # Replace by days
            if duration_string.include? "[%dd]"
              duration_string["[%dd]"] = days_left.to_s
              next
            end
            # Replace by years
            if duration_string.include? "[%yy]"
              duration_string["[%yy]"] = years.to_s
              next
            end
          end
        else
          value = nil
          case viewer
          when :seconds
            value = duration[:seconds][:total]
          when :minutes
            value = duration[:minutes][:total]
          when :hours
            value = duration[:hours][:total]
          when :days
            value = duration[:days][:total]
          when :years
            value = duration[:years][:total]
          else
            value = "No viewer format : #{viewer}"
          end
          return  value
        end
        format_date_duration.reverse!
        if options[:array]
          return format_date_duration
        end
        if format_date_duration.length > 1
        separate_format_array = format_date_duration.slice 0, format_date_duration.length - 1
        end_of_format_array = format_date_duration.slice -1
        [separate_format_array.join(", "), end_of_format_array].join(" #{@relative_format_data[:a][@lang]} ")
        else
          format_date_duration
        end
      end
    end
  end
end

